nemo/collections/asr/parts/k2/mlloss.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Copyright (c) 2020, Xiaomi CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Union
+
+import k2
+import torch
+
+from nemo.collections.asr.parts.k2.grad_utils import GradExpNormalize
+from nemo.collections.asr.parts.k2.utils import (
+    create_supervision,
+    get_tot_objf_and_finite_mask,
+    load_graph,
+    make_blank_first,
+    prep_padded_densefsavec,
+)
+
+
+class MLLoss(torch.nn.Module):
+    """
+    Maximum Likelihood criterion.
+    It is implemented as Connectionist Temporal Classification (CTC) loss,
+    but can be extended to support other loss functions (ASG, HMM, ...).
+    
+    Based on https://github.com/k2-fsa/snowfall/blob/master/snowfall/objectives/ctc.py
+    """
+
+    def __init__(
+        self,
+        num_classes: int,
+        blank: int,
+        reduction: str = "mean",
+        topo_type: str = "default",
+        topo_with_selfloops: bool = True,
+        graph_type: str = "topo",
+        aux_graph: Optional[Union[k2.Fsa, str]] = None,
```

Comment:
Like above, convert all references to `k2.*` as `'k2.*'` to prevent issues with imports of the file without K2 installed.